internal/git/gitsync_processor.go

```diff
@@ -157,8 +185,228 @@ func watchRepo(ctx context.Context, r *git.Repository, gitSync *v1.GitSync, rest
+		for _, filePatch := range patch.FilePatches() {
+			from, to := filePatch.Files()
+			if from != nil {
+				initialContent, err := getBlobFileContents(r, from)
+				if err != nil {
+					logger.Errorw("failed to get  initial content", "err", err.Error(), "repo", repo.RepoUrl)
+					return patchedResources, recentHash, err
+				}
+				err = populateResourceMap(initialContent, beforeMap, defaultNameSpace)
+				if err != nil {
+					logger.Errorw("failed to populate resource map", "err", err.Error(), "repo", repo.RepoUrl)
+					return PatchedResource{}, recentHash, err
+				}
+			}
+			if to != nil {
+				finalContent, err := getBlobFileContents(r, to)
+				if err != nil {
+					logger.Errorw("failed to get  final content", "err", err.Error(), "repo", repo.RepoUrl)
+
+					return patchedResources, recentHash, err
+				}
+				err = populateResourceMap(finalContent, afterMap, defaultNameSpace)
+				if err != nil {
+					logger.Errorw("failed to populate resource map", "err", err.Error(), "repo", repo.RepoUrl)
+					return PatchedResource{}, recentHash, err
+				}
+			}
+		}
+
+		patchedResources.Before = beforeMap
+		patchedResources.After = afterMap
+	}
+
+	return patchedResources, recentHash, nil
+}
+
+// populateResourceMap fills the resourceMap with resource names as keys and their string representations as values.
+func populateResourceMap(content []byte, resourceMap map[string]string, defaultNameSpace string) error {
+	// split the string by ---
+	resources := strings.Split(string(content), "---")
+	for _, v := range resources {
+		name, err := getResourceName(v, defaultNameSpace)
+		if err != nil {
+			return err
+		}
+		resourceMap[name] = v
+	}
+	return nil
+}
+
+// unmarshalls yaml into Kubernetes Resource
+func yamlUnmarshal(yamlContent string) (KubernetesResource, error) {
+	var resource KubernetesResource
+	err := yaml.Unmarshal([]byte(yamlContent), &resource)
+	if err != nil {
+		return KubernetesResource{}, err
+	}
+	return resource, err
+}
+
+// getResourceName extracts the name and namespace of the Kubernetes resource from YAML content.
+func getResourceName(yamlContent string, defaultNameSpace string) (string, error) {
+	resource, err := yamlUnmarshal(yamlContent)
+	if err != nil {
+		return "", err
+	}
+	namespace := resource.Metadata.Namespace
+	if namespace == "" {
+		namespace = defaultNameSpace
+	}
+	return namespace + "/" + resource.Metadata.Name, nil
+}
+
+// retrieves a specific tree (or subtree) located at a given path within a specific commit in a Git repository
+func getCommitTreeAtPath(r *git.Repository, path string, hash plumbing.Hash) (*object.Tree, error) {
+	commit, err := r.CommitObject(hash)
+	if err != nil {
+		return nil, err
+	}
+	commitTree, err := commit.Tree()
+	if err != nil {
+		return nil, err
+	}
+	commitTreeForPath, err := commitTree.Tree(path)
+	if err != nil {
+		return nil, err
+	}
+	return commitTreeForPath, nil
+}
```

Comment:
I have a comment with a "TODO" in it which says to call `getCommitTreeAtPath()` from Hao's code. Can you do that as part of this PR?

Reply:
(I noticed her logic is slightly different in that she checks for `isRootDir()` - maybe you can determine if we need that here or not